### ISSUE TITLE
Sentinel needs to support these patient_id use cases

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#3166]
There are various ways that `patient_id` gets set, we need to support them all. We currently support:

- from SMS -> new patient id, new contact
- migration -> existing **internal**[1] patient id onto new contact

We need to also support:
- [x] from SMS -> existing **external**][1] patient id onto new contact
- [x] from Enketo -> new patient id onto existing contact.  This is covered by #3000 
- [x] _using_ Enketo -> existing **external**[1] patient id onto new contact.

This ticket may have code against it, or it may just link to other tickets that take care of this.

[1] internal being a patient id we generate, external being one that we are given. At least for this ticket we are presuming that a) we are given unique patient ids, and b) they fit whatever validation scheme is configured.

**Comment by SCdF** [comment]
We will need to change sentinel so it processes person documents. We should remove the document restriction, and just update all of the existing transitions to ignore documents appropriately (some may already do so by the nature of their filters)

**Comment by abbyad** [comment]
If we were to use the same trigger, the person creation forms could be easily modified to result in a `doc.form` field. The resulting config could use this form value to assign the trigger to the doc:

```json
"registrations": [
    {
      "form": "person_create",
      "events": [
        {
          "name": "on_create",
          "trigger": "add_patient",
          "params": "",
          "bool_expr": ""
        }
      ]
    },
```

**Comment by SCdF** [comment]
Step 1: change sentinel so it passes allows transitions to transition all types of documents.

@garethbowen please review and merge https://github.com/medic/medic-sentinel/pull/111 when ready. 

I'm keeping this ticket open and working on it, but let's get this chunk out and done with.

**Comment by garethbowen** [comment]
Nice. Merged. Assigning back to you,

**Comment by SCdF** [comment]
I think adding `patient_id` to contacts created in Enketo is a separate transition, @abbyad would you agree? You don't actually expect any kind of registration action to be performed right? You'd just expect that if you created a patient contact in the UI a patient_id would be generated (unless you provided one).

**Comment by SCdF** [comment]
Actually @abbyad would you expect this to just run on any person that was created? That would generate ids for CHPs as well, right?

**Comment by SCdF** [comment]
@abbyad for the third bullet point (using Enketo -> existing external[1] patient id onto new contact.) do we need to do anything here? Can enketo forms already be configured to store something at `doc.patient_id` in those situations?

**Comment by abbyad** [comment]
> I think adding patient_id to contacts created in Enketo is a separate transition, @abbyad would you agree? You don't actually expect any kind of registration action to be performed right? You'd just expect that if you created a patient contact in the UI a patient_id would be generated (unless you provided one).

Correct. Although I don't see harm in it coming from the same transition `add_patient`, it could be a separate transition `add_patient_id` (but not that name to avoid confusion with previous trigger of same name).

> Actually @abbyad would you expect this to just run on any person that was created? That would generate ids for CHPs as well, right?

Projects that use IDs for CHWs are more likely to use an external ID, such as their "Health Worker Registry" number. Given that, I don't think we would need to create an ID for everyone in every project. If we did we might exhaust possible numbers a bit quicker, and it would mean that app workflow only projects get an unnecessary ID for patients. That may not be a problem, and may make things more consistent that every person and place has a short form ID to refer to it.

> @abbyad for the third bullet point (using Enketo -> existing external[1] patient id onto new contact.) do we need to do anything here? Can enketo forms already be configured to store something at doc.patient_id in those situations?

Yes, we can already do this. In this case we couldn't verify that it is unique, but that is likely fine.

**Comment by SCdF** [comment]
Added a WIP branch to check in code I'm working through that allows support for third party ids: https://github.com/medic/medic-sentinel/tree/3166-external-id-support

**Comment by SCdF** [comment]
@garethbowen can you please review the code: https://github.com/medic/medic-sentinel/pull/116

I have no idea where to put the following documentation. I started writing out markdown documentation to cover things, and then realised that it's sort of half-covered by `kanso.json`. Things you can specifically configure are sort of covered, but also sort of not (eg a short phrase is encouraged by the format), and there isn't a way to document things that aren't specifically tied to a static piece of configuration.

# Additional documentation

## `add_patient_id`

`add_patient_id` event has been deprecated in favour of `add_patient`

## `add_patient`

Can now be passed an externally generated patient id instead of generating one. Can be configured like so:
```json
{
  "name": "on_create",
  "trigger": "add_patient",
  "params": {
    "patient_id": "external_id"
  }
}
```

The document passed would then be expected to have a patient id in the `fields.external_id` field.

If `params.patient` exists but that field cannot be found on the document a `sys.no_provided_patient_id` error is recorded and SMS sent.
if the given external `patient_id` is not unique a `sys.provided_patient_id_not_unique` error is recorded and SMS sent.

**Comment by SCdF** [comment]
@abbyad I'd appreciate your input (or perhaps even a voice call!) on how best we can keep documentation consistent, easy to find and ideally in one place / format.

**Comment by SCdF** [comment]
(NB: code is "done" so I'm putting it up for review. However, I'd like to do some proper black-box testing and I've run out of time today, so I'm still assigning this to me and still keeping it in active work)

**Comment by garethbowen** [comment]
Some changes requested - back to you!

**Comment by abbyad** [comment]
We can start the documentation in https://github.com/medic/medic-docs/blob/master/md/config/, similar to the [sentinel transitions](https://github.com/medic/medic-docs/blob/master/md/config/transitions.md) and the [contact-summary](https://github.com/medic/medic-docs/blob/master/md/config/contact-summary.md) documentation.

**Comment by SCdF** [comment]
@abbyad OK so how does that documentation meld with `kanso.json`? Should we just ignore what is in that, presuming it is for a different situation, and replicate it in medic-docs as well?

**Comment by abbyad** [comment]
I don't think there is sufficient place to document this in `kanso.json`. We can explain "triggers", but not each one individually there. Linking to our doc from there would be useful (eg "[more info on triggers](https://github.com/medic/medic-docs/blob/master/md/config)"), as is being done [here](https://github.com/medic/medic-webapp/blob/e4330e08d9b2bd3b351b534751cf32ef1c299b04/kanso.json#L328) for contact summary.

**Comment by SCdF** [comment]
Also: https://github.com/medic/medic-webapp/pull/3260

Just some message changes, with a note to talk to @abbyad about how we should change the default config (and others??) to support auto-extending ids.

**Comment by SCdF** [comment]
OK let's get this all clear, because there are lots of fragmented things flying around.

@garethbowen, can you approve (you may have seen this stuff before):
 - Sentinel changes: https://github.com/medic/medic-sentinel/pull/116
 - Fixing some documentation / translation in Medic, including Changes.md documenation: https://github.com/medic/medic-webapp/pull/3260
 - Adding official documentation (and stubs) in the right home: https://github.com/medic/medic-docs/pull/10

Once these are cool I'm going to merge https://github.com/medic/medic-sentinel/pull/117 into https://github.com/medic/medic-sentinel/pull/116, then hopefully be able to test properly with a good local environment, and then merge into master.

**Comment by garethbowen** [comment]
@SCdF Nice one. One small change requested on each of the last two PRs. Merge when fixed!

**Comment by SCdF** [comment]
OK everything is merged with the exception of the documentation stuff, which Gareth and I will talk about on Monday, and then I'll almost certainly merge it then.

**Comment by SCdF** [comment]
OK merged documentation. Donnnnnnnnnnnne.

**Comment by ngaruko** [comment]
Any clue on how to AT this @abbyad ?

**Comment by abbyad** [comment]
Has this already been released?

**Comment by ngaruko** [comment]
Looks like it has been release @abbyad....

**Comment by SCdF** [comment]
Yes this has been released, looks like 2.11.1. However, it hasn't been AT'd ;-)
